src/rnnlm/rnnlm-lattice-rescoring.cc

```diff
@@ -0,0 +1,169 @@
+
+#include <utility>
+
+#include "rnnlm/rnnlm-lattice-rescoring.h"
+#include "util/stl-utils.h"
+#include "util/text-utils.h"
+
+namespace kaldi {
+namespace nnet3 {
+
+void KaldiRnnlmDeterministicFst::ReadFstWordSymbolTableAndRnnWordlist(
+    const std::string &rnn_wordlist,
+//    const std::string &rnn_out_wordlist,
+    const std::string &word_symbol_table_rxfilename) {
+  // Reads symbol table.
+  fst::SymbolTable *fst_word_symbols = NULL;
+  if (!(fst_word_symbols =
+      fst::SymbolTable::ReadText(word_symbol_table_rxfilename))) {
+    KALDI_ERR << "Could not read symbol table from file "
+              << word_symbol_table_rxfilename;
+  }
+
+  full_voc_size_ = fst_word_symbols->NumSymbols();
+  fst_label_to_word_.resize(full_voc_size_);
+
+  for (int32 i = 0; i < fst_label_to_word_.size(); ++i) {
+    fst_label_to_word_[i] = fst_word_symbols->Find(i);
+    if (fst_label_to_word_[i] == "") {
+      KALDI_ERR << "Could not find word for integer " << i << "in the word "
+                << "symbol table, mismatched symbol table or you have discoutinuous "
+                << "integers in your symbol table?";
+    }
+  }
+
+//  fst_label_to_rnn_out_label_.resize(fst_word_symbols->NumSymbols(), -1);
+  KALDI_LOG << "resize to " << fst_word_symbols->NumSymbols();
+  fst_label_to_rnn_label_.resize(fst_word_symbols->NumSymbols(), -1);
+
+  out_OOS_index_ = 1;
+  {
+    std::ifstream ifile(rnn_wordlist.c_str());
+    int32 id;
+    string word;
+    int32 i = 0;
+    while (ifile >> word >> id) {
+      if (word == "</s>") {
+        final_word_index_ = id;
+      }
+      KALDI_ASSERT(i == id);
+      i++;
+      rnn_label_to_word_.push_back(word);
+
+//      if (word == "<brk>") {
+//        continue;
+//      }
+
+      int fst_label = fst_word_symbols->Find(rnn_label_to_word_[id]);
+      if (fst::SymbolTable::kNoSymbol != fst_label || id == out_OOS_index_ || id == 0 || word == "<brk>") {}
+      else {
+        KALDI_LOG << "warning: word " << word << " in RNNLM wordlist but not in FST wordlist";
+      }
+      if (id != out_OOS_index_ && out_OOS_index_ != 0 && fst_label != -1) {
+        fst_label_to_rnn_label_[fst_label] = id;
+      }
+    }
+  }
+
+  for (int32 i = 0; i < fst_label_to_rnn_label_.size(); i++) {
+    if (fst_label_to_rnn_label_[i] == -1) {
+      fst_label_to_rnn_label_[i] = out_OOS_index_;
+    }
+  }
+}
+
+KaldiRnnlmDeterministicFst::KaldiRnnlmDeterministicFst(int32 max_ngram_order,
+    const std::string &rnn_wordlist,
+    const std::string &word_symbol_table_rxfilename,
+    const DecodableRnnlmSimpleLoopedInfo &info) {
+  max_ngram_order_ = max_ngram_order;
+  ReadFstWordSymbolTableAndRnnWordlist(rnn_wordlist,
+                                       word_symbol_table_rxfilename);
+
+  std::vector<Label> bos;
+  bos.push_back(0); // 0 for <s>
+  state_to_wseq_.push_back(bos);
+  DecodableRnnlmSimpleLooped decodable_rnnlm(info);
+  decodable_rnnlm.TakeFeatures(std::vector<Label>(1, bos[0]));
```

Comment:
It looks to me like you are just calling TakeFeatures with a copy of the 'bos' vector, constructed in a rather strange way.  Why not use bos (rename this to bos_seq) directly?